### PR TITLE
[SHIP-540] python generator plugin interfaces

### PR DESCRIPTION
**Pull request by dkolas**
Initial generator plugin interface, with test generator which reverses text.

Depends on engine changes which are not yet deployed to prod.

Review request to especially make sure the parameters of `GenerateRequest` make sense and are sufficiently commented.

### CODE REVIEW

File: src/steamship/data/operations/generator.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
+from __future__ import annotations
+
+from typing import List, Optional
+
+from steamship.base.request import Request
+from steamship.base.response import Response
+from steamship.data.block import Block
+
+
+class GenerateRequest(Request):
```
Comment (douglas-reid) [inline src/steamship/data/operations/generator.py:10]: any reason not to include a `kwargs` style extra bit here?  I feel like we need a way to pass runtime args to calls to generate.

Comment (douglas-reid) [reply src/steamship/data/operations/generator.py:10]: the text example was "stop" keywords

Comment (douglas-reid) [reply src/steamship/data/operations/generator.py:10]: NOTE TO FUTURE SELF: we will address side-channel info in a separate PR.

File: src/steamship/data/operations/generator.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    """This class provides the input for a request to a Generator.  There are several ways to specify the input; see below"""
+
+    # Input Specification
+
+    # Must specify plugin instance to use
+    plugin_instance: str = None
```
Comment (douglas-reid) [inline src/steamship/data/operations/generator.py:16]: clarity: this is instance handle, correct?

**Review by douglas-reid — COMMENTED** [review]
Just some questions re: Request surface.

File: src/steamship/data/operations/generator.py (hunk 2 above)
Comment (dkolas) [reply src/steamship/data/operations/generator.py:16]: Yeah, I copied the behavior of tag, which is awkward. Can change here if the inconsistency is better than the bad-naming

File: src/steamship/data/operations/generator.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+    # May specify blocks by their file_id. If so, may specify start and end index
+    input_file_id: str = None
```
[the remark this answers is not in the document]
Comment (dkolas) [reply src/steamship/data/operations/generator.py:19]: I think that would be a lovely add-on

File: src/steamship/data/operations/generator.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    input_file_start_block_index: int = None
+    input_file_end_block_index: Optional[
+        int
+    ] = None  # EXCLUSIVE end index, like most programming languages
+
+    # May specify raw text
+    text: Optional[str] = None
```
[the remark this answers is not in the document]
Comment (dkolas) [reply src/steamship/data/operations/generator.py:26]: I can add comments to make this more explicit.

File: src/steamship/data/operations/generator.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+    # Desired output specification
+
+    # Whether we want the output appended to a file
+    append_output_to_file: bool = False
+
+    # May specify a file to which to append the results.  This may be the same file as
+    # the input or not.  If appendOutputToFile is true but the outputFileId is not set,
+    # create a new file.
+    output_file_id: Optional[str] = None
```
[the remark this answers is not in the document]
Comment (dkolas) [reply src/steamship/data/operations/generator.py:45]: Yes,  should add-on

**Pull request merged**